blog/_src/posts/2019-05-14-scoping-in-r.md

```diff
@@ -0,0 +1,329 @@
+defined? Or is it bound to the `x` defined on line 4, in the same dynamic
+environment that `f` was called?
+
+
+### Lexical scoping
+
+Under _lexical scoping_ (also known as _static scoping_), a free variable is
+evaluated in its _lexical environment_, i.e., where it is defined. In the example
+above, the free variable `x` is used in the function `f`, which is defined at
+the top-level environment. At that point in the program, there is a definition
+of `x` visible, with value `1`.
+
+Therefore, under lexical scoping, the example program returns `1`.
+
+Most programming languages we use today are lexically scoped. In principle, a
+compiler could statically determine what values free variables are bound to.
+
+
+### Dynamic scoping
+
+Under _dynamic scoping_, a free variable is evaluated in its _dynamic
+environment_, i.e. where it is called. In other words, the variable is bound to
+the value last assigned to it (at run time). In our example, just before `f` is
+called, `x` is assigned the value `2`.
+
+Therefore, under dynamic scoping, the example program returns `2`.
+
+Dynamically scoped programming languages include bash and LaTeX. Emacs Lisp is
+dynamically scoped, but allows the programmer to opt into lexical scoping.
+Conversely, Perl and Common Lisp are lexically scoped by default but allow the
+programmer to choose dynamic scoping if desired.
+
+
+## What about R?
+
+The example is written in valid R. If you run it, you will see that the result
+is `1`. Therefore, according to our litmus test, **R is lexically scoped**. But
+is this the whole story?
+
+In _[Evaluating the Design of the R Language][mhov2012]_ (ECOOP 2012)
+([doi][mhov2012-doi]), Morandat, Hill, Osvald, and Vitek write:
+
+> As is often the case, R is lexically scoped up to the point it is not. R
+> is above all a dynamic language with full reflective access to the running
+> program’s data and representation.
+
+[mhov2012]: http://janvitek.org/pubs/ecoop12.pdf
+[mhov2012-doi]: https://doi.org/10.1007/978-3-642-31057-7_6
+
+
+### Circumventing scoping in R
+
+There are many ways to "defeat" scoping in R---made easier because it supports
+explicit manipulation of environments.
+
+Here are some different tricks we can try.
+
+#### Evaluating arbitrary code
+
+```r
+x <- 1
+f <- function(a) {
+  eval(parse(text = "x <- 0"))
+  x + a
+}
+g <- function() {
+  x <- 2
+  f(0)
+}
+g() # 0
+```
+
+On line 3, we parse the string `"x <- 0"` into an unevaluated R expression and
+then evaluate it. This assigns `0` to `x`, which is no longer a free variable
+when it is used on line 4.
+
+#### Simulating dynamic scope
+
+```r
+x <- 1
+f <- function(a) {
+  get("x", envir = parent.frame()) + a
+}
+g <- function() {
+  x <- 2
+  f(0)
+}
+g() # 2
+```
+
+On line 3, we perform an explicit variable lookup for `x`, but we do so in the
+environment `parent.frame()`, which refers to the calling function's
+environment.
+
+#### Constructing an arbitrary environment
```

Comment:
I'm not sure the example in this section brings much in comparison with just `x <- 3` inside `f`. It'd make more sense if there were some plot twist hiding later, which references this section, but there is no, from what I can see. Maybe in the next section, you could say that “we can remove as well as add (cf. prev. section)”.

Reply:
I wanted this to be an example of explicit environment manipulation. It's slightly different from adding `x <- 3` inside the function, because every other occurrence of `x` in the function will have the value `1`.

I'll see if I can make that point more explicit.